src/hashtable.c

```diff
@@ -0,0 +1,163 @@
+ * tightly together in their own key table and avoid referencing the element
+ * table and elements as much as possible. Key value zero is reserved as a
+ * marker for an empty bucket to avoid checking for NULL in the element table.
+ * If we do get a hash value of zero, we -1 to wrap it around to 0xffff. */
+
+/* Use max 0.8 load factor to avoid bad open addressing performance. */
+#define HASHTABLE_LOADFACTOR_NUM 8
+#define HASHTABLE_LOADFACTOR_DEN 10
+
+hashtable_t *hashtable_new(int size, hash_f hash, cmp_f cmp)
+{
+    hashtable_t *t;
+    int size2;
+
+    /* Adjust requested size to account for max load factor. */
+    size = 1 + size * HASHTABLE_LOADFACTOR_DEN / HASHTABLE_LOADFACTOR_NUM;
+    /* Use next power of 2 larger than the requested size. */
+    for (size2 = 1; size2 < size; size2 <<= 1) ;
+    if (!(t = calloc(1, sizeof(hashtable_t)+ size2 * sizeof(unsigned))))
+        return NULL;
+    if (!(t->etable = calloc(size2, sizeof(void *)))) {
+        free(t);
+        return NULL;
+    }
+    t->size = size2;
+    t->count = 0;
+    t->hash = hash;
+    t->cmp = cmp;
+#ifndef HASHTABLE_NSTATS
+    t->find_count = t->match_count = t->keycmp_count = t->entrycmp_count = 0;
+#endif
+    return t;
+}
+
+void hashtable_free(hashtable_t *t)
+{
+    if (t) {
+        free(t->etable);
+        free(t);
+    }
+}
+
+/* MurmurHash3 finalization mix function. */
+static inline unsigned mix32(unsigned int h)
+{
+    h ^= h >> 16;
+    h *= 0x85ebca6b;
+    h ^= h >> 13;
+    h *= 0xc2b2ae35;
+    h ^= h >> 16;
+    return h;
+}
+
+/* Get hash key, reserving zero as an empty bucket marker. */
+static inline unsigned get_key(const hashtable_t *t, const void *e)
+{
+    unsigned k = t->hash(e);
+    return k ? k : -1;
+}
+
+/* Prefix macro for probing table t for entry e with key k and index i. */
+#define do_probe(t, e, k) \
+    const unsigned k = get_key(t, e);\
+    const unsigned mask = t->size - 1;\
+    const unsigned index = mix32(k) & mask;\
+    unsigned i = index, s = 0;\
+    do
+
+/* Suffix macro for do_probe. */
+#define while_probe \
+    while ((i = (i + ++s) & mask) != index)
+
+void *hashtable_add(hashtable_t *t, void *e)
+{
+    assert(e != NULL);
+    do_probe(t, e, k) {
+        if (!t->ktable[i]) {
+            t->count++;
+            t->ktable[i] = k;
+            return t->etable[i] = e;
+        }
+    } while_probe;
+    return NULL;
+}
+
+/* Conditional macro for incrementing stats counters. */
+#ifndef HASHTABLE_NSTATS
+#define stats_inc(c) (c++)
+#else
+#define stats_inc(c)
+#endif
+
+void *hashtable_find(hashtable_t *t, void *m)
+{
+    assert(m != NULL);
+    void *e;
+    unsigned ke;
+
+    stats_inc(t->find_count);
+    do_probe(t, m, km) {
+        if (!(ke = t->ktable[i]))
+            return NULL;
+        stats_inc(t->keycmp_count);
+        if (km == ke) {
+            stats_inc(t->entrycmp_count);
+            if (!t->cmp(m, e = t->etable[i])) {
+                stats_inc(t->match_count);
```

Comment:
I found them very useful for getting a feel for what the search was doing and where the costs were. There is also support for turning them of by defining HASHTABLE_NSTATS, and all those stats_inc() calls get #defined to nothing (this also turns off the match stats support in sumset.c and rdiff).

I also found that there was no measurable difference when turning them off, so I've got them defaulting to on. We could add a CMake option to turn them off.